File: python/remove-nth-from-end.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        right = head
        left = head
        i = 0
        while n > i:
            right = right.next
            n = n-1
        if(right == None):
            return head.next
        while right.next:
            right = right.next
            left = left.next
        left.next = left.next.next
        return head

File: java/remove-nth-from-end.java
class Solution {
  public ListNode removeNthFromEnd(ListNode head, int n) {
    // 1 <= n <= sz
    ListNode right = head;
    for (int i = 0; i < n; i++)
      right = right.next;
    if (right == null)
      return head.next;
    ListNode left = head;
    while (right.next != null) {
      right = right.next;
      left = left.next;
    }
    left.next = left.next.next;
    return head;
  }
}
